### PR TITLE
feat(envelope): Add some useful envelope methods

### PR DESCRIPTION
**Pull request by dashed**
Add some useful envelope methods needed for integration tests in https://github.com/getsentry/relay/pull/724

## TODO

- [x] add some tests

### CODE REVIEW

File: sentry_sdk/envelope.py [hunk 1]
```diff
@@ -71,6 +71,12 @@ def add_item(
         # type: (...) -> None
         self.items.append(item)
 
+    def get_items(
+        self, type=None  # type: Optional[str]
+    ):
+        # type: (...) -> List[Item]
+        return [item for item in self.items if item.get_type() == type]
```
[the remark this answers is not in the document]
Comment (untitaker) [reply sentry_sdk/envelope.py:78]: That doesn't seem like sensible behavior to me for a call like `envelope.get_items()`

[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py:78]: @untitaker I'm open to suggestions.

[the remark this answers is not in the document]
Comment (untitaker) [reply sentry_sdk/envelope.py:78]: yes so either making type non-optional or writing the condition like rodolfo suggested (return `self.items` if type is None) is fine by me. Do we rely on this behavior anywhere?

[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py:81]: Fixed in https://github.com/getsentry/sentry-python/pull/793/commits/6dbd284a76ccb1364601c94b9939434677da32ae

File: sentry_sdk/envelope.py [hunk 2]
```diff
@@ -79,6 +87,14 @@ def get_event(self):
                 return event
         return None
 
+    def get_transaction_event(self):
+        # type: (...) -> Optional[Event]
+        for items in self.items:
+            event = items.get_transaction_event()
```
Comment (markstory) [inline sentry_sdk/envelope.py:93]:
```suggestion
        for item in self.items:
            event = item.get_transaction_event()
```

File: sentry_sdk/envelope.py [hunk 3]
```diff
@@ -279,7 +305,7 @@ def deserialize_from(
         headers = json.loads(line)
         length = headers["length"]
         payload = f.read(length)
-        if headers.get("type") == "event":
+        if headers.get("type") == "event" or headers.get("type") == "transaction":
```
Comment (rhcarvalho) [inline sentry_sdk/envelope.py:308]:
```suggestion
        if headers.get("type") in ("event", "transaction"):
```

**Review by rhcarvalho — COMMENTED** [review]
@dashed I see you want this for https://github.com/getsentry/relay/pull/724, but since the code is not used neither here nor in the Relay PR, it is hard to judge whether it actually helps writing what you're trying to write ;)

One suggestion would be to add simple tests here in sentry-python to showcase the usage (I would not expect someone maintaining this repo to have to read Relay tests).

```py
# t = ...
e = Envelope()
e.add_transaction(t)
with pytest.raises(ValueError):
    e.add_transaction(t)  # can only have 1 transaction in the envelope
assert e.get_transaction() is t
```

Since there can only be one transaction, perhaps a better API for the present would be a single property -- with the drawback of getting ugly if we ever allow more than 1 transaction per envelope. If that ever happens we can deprecate the property and add `add_transaction`, perhaps not a big deal.

```py
# t = ...
e = Envelope()
t.transaction = t  # behind the scenes this either adds or replaces an envelope item
assert t.transaction == t
```

File: sentry_sdk/envelope.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py:80]: I'll remove this function. Originally this was for fetching `measurement` items.

File: sentry_sdk/envelope.py [hunk 4]
```diff
@@ -250,6 +270,12 @@ def get_event(self):
             return self.payload.json
         return None
 
+    def get_transaction_event(self):
+        # type: (...) -> Optional[Event]
+        if self.headers.get("type") == "transaction" and self.payload.json is not None:
```
[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py]: Done in https://github.com/getsentry/sentry-python/pull/793/commits/bfd3ff28f0e01e526b4ce1b175ac67c588f0e2b5

File: sentry_sdk/envelope.py [hunk 5]
```diff
@@ -227,6 +243,10 @@ def __repr__(self):
             self.data_category,
         )
 
+    def get_type(self):
+        # type: (...) -> Optional[str]
+        return self.headers.get("type")
```
[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py:241]: Done in https://github.com/getsentry/sentry-python/pull/793/commits/bfd3ff28f0e01e526b4ce1b175ac67c588f0e2b5

File: sentry_sdk/envelope.py [hunk 6]
```diff
@@ -79,6 +87,14 @@ def get_event(self):
                 return event
         return None
 
+    def get_transaction_event(self):
+        # type: (...) -> Optional[Event]
+        for items in self.items:
```
[the remark this answers is not in the document]
Comment (dashed) [reply sentry_sdk/envelope.py]: I removed `get_items()`.

It should be the first transaction that'll be processed. This is the behaviour that Relay expects today.

**Comment by dashed** [comment]
@rhcarvalho I've added some tests https://github.com/getsentry/sentry-python/pull/793/commits/1e29b2ab0a03070ae7ad622ff84f29007023a2dd

**Review by untitaker — APPROVED** [review]

**Pull request merged**